File: backend/src/main/java/com/hex/projectgovern/module/risk/Risk.java
package com.hex.projectgovern.module.risk;

import com.hex.projectgovern.common.entity.SoftDeletableEntity;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;

/**
 * 项目风险登记(Risk Register)。
 * <p>对齐 V2.6 {@code risk} 表。
 * <p>关键设计:
 * <ul>
 *   <li>score = probability × impact (1-25), level 由 score 自动推导 (LOW/MEDIUM/HIGH/CRITICAL)</li>
 *   <li>probability / impact 都是 1-5 的离散值, 渲染 5x5 风险矩阵方便</li>
 *   <li>软关联 WBS 任务 / 里程碑(可选, 风险不一定挂在具体任务上)</li>
 *   <li>状态机: OPEN → MITIGATING → CLOSED, 还可以 OCCURRED(已发生) 或 ACCEPTED(接受)</li>
 * </ul>
 */
@Entity
@Table(name = "risk", uniqueConstraints = {
        @UniqueConstraint(name = "uk_risk_project_code", columnNames = {"project_id", "code"})
})
@Getter @Setter @NoArgsConstructor
public class Risk extends SoftDeletableEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "project_id", nullable = false)
    private Long projectId;

    @Column(nullable = false, length = 32)
    private String code;          // R-001, R-002 ...

    @Column(nullable = false, length = 256)
    private String title;

    @Column(columnDefinition = "text")
    private String description;

    @Column(nullable = false, length = 16)
    private String category;       // TECHNICAL / SCHEDULE / COST / QUALITY / EXTERNAL / ORGANIZATIONAL / OTHER

    @Column(nullable = false)
    private Integer probability;   // 1-5

    @Column(nullable = false)
    private Integer impact;        // 1-5

    @Column(nullable = false)
    private Integer score;         // 1-25 (probability × impact, 自动算)

    @Column(nullable = false, length = 16)
    private String level;          // LOW / MEDIUM / HIGH / CRITICAL (由 score 自动算)

    @Column(nullable = false, length = 16)
    private String status = "OPEN";  // OPEN / MITIGATING / CLOSED / OCCURRED / ACCEPTED

    @Column(name = "owner_user_id")
    private Long ownerUserId;

    @Column(columnDefinition = "text")
    private String mitigation;        // 预防/缓解措施

    @Column(columnDefinition = "text")
    private String contingency;       // 应急/兜底措施

    @Column(name = "response_strategy", length = 16)
    private String responseStrategy;  // AVOID / MITIGATE / TRANSFER / ACCEPT / EXPLOIT / ENHANCE / SHARE

    @Column(name = "identified_date", nullable = false)
    private LocalDate identifiedDate = LocalDate.now();

    @Column(name = "target_close_date")
    private LocalDate targetCloseDate;

    @Column(name = "actual_close_date")
    private LocalDate actualCloseDate;

    @Column(name = "related_wbs_task_id")
    private Long relatedWbsTaskId;

    @Column(name = "related_milestone_id")
    private Long relatedMilestoneId;

    @Column(name = "created_by")
    private Long createdBy;

    // ============================================================
    // 静态工具: score → level 推导
    // ============================================================
    public static int computeScore(int probability, int impact) {
        return Math.max(1, Math.min(25, probability * impact));
    }

    public static String levelOf(int score) {
        if (score >= 16) return "CRITICAL";
        if (score >= 10) return "HIGH";
        if (score >= 5)  return "MEDIUM";
        return "LOW";
    }

    /** 配置版: 从 SystemConfigService 读阈值,fallback 到静态默认值 */
    public static String levelOf(int score, com.hex.projectgovern.module.admin.SystemConfigService cfg) {
        int crit = cfg.getInt("business.risk.score_threshold.critical", 16);
        int high = cfg.getInt("business.risk.score_threshold.high", 10);
        int med  = Math.min(high - 1, 5);
        if (score >= crit) return "CRITICAL";
        if (score >= high) return "HIGH";
        if (score >= med)  return "MEDIUM";
        return "LOW";
    }

    /** 同时刷新 score + level — 修改 probability/impact 后必调 */
    public void recomputeScoreAndLevel() {
        recomputeScoreAndLevel(null);
    }

    /** 配置版: 传入 SystemConfigService 走 UI 阈值,传 null 走静态默认 */
    public void recomputeScoreAndLevel(com.hex.projectgovern.module.admin.SystemConfigService cfg) {
        if (probability == null || impact == null) return;
        this.score = computeScore(probability, impact);
        this.level = cfg == null ? levelOf(score) : levelOf(score, cfg);
    }
}

File: frontend/src/api/risk.ts
/**
 * Risks API (P4 风险管理)
 *
 * 后端: /api/risks/**
 *  - 风险主表 / 应对行动 / 变更历史 / 5x5 矩阵 / 健康度 KPI
 *  - client.ts 拦截器已解 ApiResponse 包装, 这里直接返回 T
 */
import api from './client'

/** 风险主表 — 对齐后端 RiskResponse */
export interface RiskItem {
  id: number
  projectId: number
  title: string
  description?: string
  category?: string
  level: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW'
  status: 'OPEN' | 'MITIGATING' | 'OCCURRED' | 'ACCEPTED' | 'CLOSED'
  // ... 其余字段
  [k: string]: any
}

// 兼容旧组件引用
export type RiskLevel = RiskItem['level']
export type RiskStatus = RiskItem['status']

/** 风险主表 — 对齐后端 RiskResponse (详细字段) */
export interface RiskItemFull {
  id: number
  projectId: number
  code: string // R-001
  title: string
  description: string | null
  category: 'TECHNICAL' | 'SCHEDULE' | 'COST' | 'QUALITY' | 'EXTERNAL' | 'ORGANIZATIONAL' | 'OTHER'
  probability: number // 1-5
  impact: number // 1-5
  score: number // 1-25 (probability × impact)
  level: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'
  status: 'OPEN' | 'MITIGATING' | 'CLOSED' | 'OCCURRED' | 'ACCEPTED'
  ownerUserId: number | null
  ownerName: string | null
  mitigation: string | null
  contingency: string | null
  responseStrategy: 'AVOID' | 'MITIGATE' | 'TRANSFER' | 'ACCEPT' | 'EXPLOIT' | 'ENHANCE' | 'SHARE' | null
  identifiedDate: string | null
  targetCloseDate: string | null
  actualCloseDate: string | null
  relatedWbsTaskId: number | null
  relatedWbsTaskName: string | null
  relatedMilestoneId: number | null
  relatedMilestoneName: string | null
  createdBy: number | null
  createdAt: string
  updatedAt: string
}

/** 风险主表 — 创建/更新请求 (id 缺失=新建) */
export interface RiskRequest {
  id?: number
  projectId: number
  code: string
  title: string
  description?: string | null
  category: string
  probability: number
  impact: number
  status?: string
  ownerUserId?: number | null
  mitigation?: string | null
  contingency?: string | null
  responseStrategy?: string | null
  identifiedDate?: string | null
  targetCloseDate?: string | null
  relatedWbsTaskId?: number | null
  relatedMilestoneId?: number | null
}

// ============================================================
// 应对行动
// ============================================================

export interface RiskResponseItem {
  id: number
  riskId: number
  action: string
  ownerUserId: number | null
  ownerName: string | null
  dueDate: string | null
  completedAt: string | null
  status: 'PLANNED' | 'IN_PROGRESS' | 'DONE' | 'CANCELLED'
  note: string | null
  createdAt: string
}

export interface RiskResponseRequest {
  id?: number
  action: string
  ownerUserId?: number | null
  dueDate?: string | null
  status?: string
  note?: string | null
}

// ============================================================
// 变更历史
// ============================================================

export interface RiskHistoryItem {
  id: number
  riskId: number
  action:
    | 'CREATED'
    | 'STATUS_CHANGED'
    | 'SCORE_CHANGED'
    | 'OWNER_CHANGED'
    | 'LEVEL_CHANGED'
    | 'COMMENTED'
    | 'RESPONSE_ADDED'
    | 'RESPONSE_DONE'
    | 'DELETED' // V2.7 新增, 软删专用
  fieldName: string | null
  oldValue: string | null
  newValue: string | null
  comment: string | null
  operatorId: number | null
  operatorName: string | null
  createdAt: string
}

// ============================================================
// 健康度 + 矩阵
// ============================================================

export interface RiskHealthSummary {
  projectId: number
  totalCount: number
  activeCount: number
  criticalActive: number
  highActive: number
  occurredCount: number
  maxActiveScore: number
  byCategory: Record<string, number>
  byLevel: Record<string, number>
}

export interface RiskMatrixCell {
  probability: number
  impact: number
  count: number
  risks: RiskItem[]
}

export interface RiskMatrix {
  cells: RiskMatrixCell[]
}

// ============================================================
// API 调用
// ============================================================

/** 拉项目全部风险 (按 score 降序, 含已关闭) */
export function getRisksByProject(projectId: number) {
  return api.get<RiskItem[]>(`/risks/by-project/${projectId}`)
}

/** 拉项目活跃风险 (排除 CLOSED/ACCEPTED) */
export function getActiveRisks(projectId: number) {
  return api.get<RiskItem[]>(`/risks/by-project/${projectId}/active`)
}

/** 单条 */
export function getRisk(id: number) {
  return api.get<RiskItem>(`/risks/${id}`)
}

/** 新建/更新 */
export function saveRisk(req: RiskRequest) {
  return api.post<RiskItem>('/risks', req)
}

/** 软删 */
export function deleteRisk(id: number) {
  return api.delete<void>(`/risks/${id}`)
}

/** 应对行动 */
export function getRiskResponses(riskId: number) {
  return api.get<RiskResponseItem[]>(`/risks/${riskId}/responses`)
}
export function saveRiskResponse(riskId: number, req: RiskResponseRequest) {
  return api.post<RiskResponseItem>(`/risks/${riskId}/responses`, req)
}
export function deleteRiskResponse(responseId: number) {
  return api.delete<void>(`/risks/responses/${responseId}`)
}

/** 历史 */
export function getRiskHistory(riskId: number) {
  return api.get<RiskHistoryItem[]>(`/risks/${riskId}/history`)
}

/** 健康度 */
export function getRiskHealth(projectId: number) {
  return api.get<RiskHealthSummary>(`/risks/health/by-project/${projectId}`)
}

/** 5x5 矩阵 */
export function getRiskMatrix(projectId: number) {
  return api.get<RiskMatrix>(`/risks/matrix/by-project/${projectId}`)
}
